### PR TITLE
[Serve] Deflake `test_metrics`

### LINKED ISSUE #45843: CI test windows://python/ray/serve/tests:test_metrics is consistently_failing
CI test **windows://python/ray/serve/tests:test_metrics** is flaky. Recent failures: 
	- https://buildkite.com/ray-project/postmerge/builds/4811#0190008e-b34d-496e-a786-739c01c7f82a
	- https://buildkite.com/ray-project/postmerge/builds/4652#018fc58f-1bca-42f0-8046-2660cd83c333
	- https://buildkite.com/ray-project/postmerge/builds/4530#018f96cc-34fa-421d-b878-313829bf274b

DataCaseName-windows://python/ray/serve/tests:test_metrics-END
Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by GeneDer**
## Why are these changes needed?

Split out test_metrics to run on it's own so the metrics will not be polluted by other tests. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/45843

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by GeneDer** [comment]
Tests 5 runs ([29452](https://buildkite.com/ray-project/premerge/builds/29452#0192109f-97f5-46da-b89e-c552bfc3d1f6/6-11219), [29455](https://buildkite.com/ray-project/premerge/builds/29455#01921103-6e60-46a5-8d7c-05586028a0e7/6-11255), [29457](https://buildkite.com/ray-project/premerge/builds/29457#0192114f-aa14-405c-a664-d9103391f891/6-11231), [29459](https://buildkite.com/ray-project/premerge/builds/29459#0192118b-2497-4b83-9035-b93019f02a2d/6-11222), [29461](https://buildkite.com/ray-project/premerge/builds/29461#019211cf-d078-4eef-ab2b-7038b021400e/6-11216)) and not seeing any more failures with this change

**Review by zcin — APPROVED** [review]

File: python/ray/serve/tests/test_metrics.py [hunk 1]
```diff
@@ -32,7 +33,7 @@
 TEST_METRICS_EXPORT_PORT = 9999
 
 
-@pytest.fixture
+@pytest_asyncio.fixture
```
Comment (edoakes) [inline python/ray/serve/tests/test_metrics.py:36]: I think we usually use `pytest.mark.asyncio`

Comment (GeneDer) [reply python/ray/serve/tests/test_metrics.py:36]: Yea, I think `pytest.fixture` getting deprecated in some future version. Saw this warning message in one of the failed run https://buildkite.com/ray-project/premerge/builds/29448#0192100b-ce04-4360-bcb2-a33f09cc1442/6-8202. But I can revert if we want to follow up to clean all the usage of `pytest.fixture` when that comes

**Comment by GeneDer** [comment]
> @GeneDer can you explain a little more about what the issue was an how this solves it? And also why is this specific to windows?

This test was running in windows before, there is no change on that. 

I think the main issue is there are some race conditions between different tests running in this build and polluting the metrics, so often time we see this test failing (timed out with condition never met). 

Also after I factor this test out, we are seeing those [unexpected kwarg error](https://buildkite.com/ray-project/premerge/builds/29448#0192100b-ce04-4360-bcb2-a33f09cc1442/6-7924), so those get fixed as well.

**Comment by edoakes** [comment]
> > @GeneDer can you explain a little more about what the issue was an how this solves it? And also why is this specific to windows?
> 
> This test was running in windows before, there is no change on that.
> 
> I think the main issue is there are some race conditions between different tests running in this build and polluting the metrics, so often time we see this test failing (timed out with condition never met).
> 
> Also after I factor this test out, we are seeing those [unexpected kwarg error](https://buildkite.com/ray-project/premerge/builds/29448#0192100b-ce04-4360-bcb2-a33f09cc1442/6-7924), so those get fixed as well.

We should be able to set up the fixtures properly to clear all of the expected metrics. We rely on metrics in quite a few tests, so it's probably better to nail down the root cause there

**Comment by GeneDer** [comment]
@edoakes I refactored the call to clean up metrics between tests into fixture and call on each of the tests here. Can you PTAL?

File: python/ray/serve/tests/test_metrics.py [hunk 2]
```diff
@@ -32,6 +32,21 @@
...
+    requests.post(delete_all_series_url)
+    requests.post(clean_tombstones_url)
+    yield
+    requests.post(delete_all_series_url)
+    requests.post(clean_tombstones_url)
+
+
 @pytest.fixture
 def serve_start_shutdown():
     """Fixture provides a fresh Ray cluster to prevent metrics state sharing."""
```
Comment (edoakes) [inline python/ray/serve/tests/test_metrics.py:52]: Hm shouldn't this be sufficient on its own? The prometheus endpoint is the raylet, so if ray is shut down between runs there should be no state sharing.

What am I missing?

Comment (GeneDer) [reply python/ray/serve/tests/test_metrics.py:52]: My expectation is there are something that's not cleaned up in between those tests. And in fact adding those calls seems to helped. Now that thinking through it again maybe just adding some sleep in between will also help the same way and maybe the issue is serve and/or ray wasn't complete shutdown before the next test starts? 🤔 Let me do some more experiments

Comment (edoakes) [reply python/ray/serve/tests/test_metrics.py:52]: we should not add any sleeps -- if we need to wait for anything to clean up, then explicitly wait for the cleanup to happen

sleeps are what make things flaky in the first place

**Comment by GeneDer** [comment]
Close for now, will dig into this deeper when we have more bandwidth. There still seemed to have some port binding issues with this change

**Pull request closed without merging**